xwiki-commons-core/xwiki-commons-component/xwiki-commons-component-default/src/main/java/org/xwiki/component/embed/EmbeddableComponentManager.java

```diff
@@ -231,33 +233,43 @@ public <T> List<T> getInstanceList(Type role) throws ComponentLookupException
     @SuppressWarnings("unchecked")
     public <T> Map<String, T> getInstanceMap(Type roleType) throws ComponentLookupException
     {
-        Map<String, T> components = new HashMap<>();
-
-        Map<String, ComponentEntry<?>> entries = this.componentEntries.get(roleType);
-
-        // Add local components
-        if (entries != null) {
-            for (Map.Entry<String, ComponentEntry<?>> entry : entries.entrySet()) {
+        Map<String, ComponentEntry<?>> entries = new HashMap<>();
+        if (this.componentEntries.containsKey(roleType)) {
+            entries.putAll(this.componentEntries.get(roleType));
+            for (ComponentEntry<?> value : entries.values()) {
                 try {
-                    components.put(entry.getKey(), getComponentInstance((ComponentEntry<T>) entry.getValue()));
+                    getComponentInstance(value); // force loading the instance
                 } catch (Exception e) {
                     throw new ComponentLookupException(
-                        "Failed to lookup component with type [" + roleType + "] and hint [" + entry.getKey() + "]", e);
+                        String.format("Failed to lookup component with type [%s] and hint [%s]",
+                            roleType, value.descriptor.getRoleHint())
+                        , e);
                 }
             }
         }
 
         // Add parent components
         if (getParent() != null) {
-            // If the hint already exists in the children Component Manager then don't add the one from the parent.
-            for (Map.Entry<String, T> entry : getParent().<T>getInstanceMap(roleType).entrySet()) {
-                if (!components.containsKey(entry.getKey())) {
-                    components.put(entry.getKey(), entry.getValue());
+            for (ComponentDescriptor<T> componentDescriptor : getParent().<T>getComponentDescriptorList(roleType)) {
+                String roleHint = componentDescriptor.getRoleHint();
+                if (!entries.containsKey(roleHint)) {
+                    T instance = getParent().getInstance(roleType, roleHint);
+                    entries.put(roleHint, new ComponentEntry<>(componentDescriptor, instance));
                 }
             }
         }
 
-        return components;
+        // Return a map ordered by the role type priority
+        return entries.entrySet().stream()
```

Comment:
@tmortagne I'm not super happy with this solution (hence the draft PR) because it means to compute back the ordering at each call. I'm a bit afraid about the perf impact, now I'm not sure how to do it better knowing that even if we order the components in the current manager, we'll need to reorder here because of the other components retrieved from the parents. wdyt?

Reply:
I'm not sure we really need to order the Map. Shouldn't the ordering be done only in getInstanceList ?

Reply:
Well the advantage to do it there, is that we automatically have the list ordered since it follows the iteration order of the map. Now if we decide to only have the ordering on the list I'm not sure it would be a lot better in perf. The only advantage I can see is that the code would be a bit simpler. 
Note that I haven't ordered the `#getComponentDescriptorList` either.

Reply:
Yeah, I guess it does not hurt.

>  Note that I haven't ordered the #getComponentDescriptorList either.

Would be more consistent to apply the ordering there too since the order is provider by the descriptor.

Reply:
> Would be more consistent to apply the ordering there too since the order is provided by the descriptor.

Yes I was thinking about doing it, but I wanted to be sure that this first implementation was ok. So you don't think we could have strong performance issue here?

Reply:
Ordering a few components is not a big issue IMO compared to the whole process of gathering those components. Especially since getComponentList is generally not used in a CPU intensive way.

I'm more concerned about the impact of the role hint priority support which, in theory, needs to select the component with the higher priority among all the members of the ComponentManager hierarchy before (initializing it if it's not already the case and) returning it, which is a big difference with current behavior. Now in theory pretty much mean one or two more Map#get, so the impact should not be that huge.

Reply:
> I'm more concerned about the impact of the role hint priority support which, in theory, needs to select the component descriptor with the highest priority among all the members of the ComponentManager hierarchy before (initializing it if it's not already the case and) returning it, which is a big difference with current behavior. Now in theory, it pretty much mean one or two more Map#get, so the impact should not be that huge.

Not sure to follow you there, have looked at my changes on `ComponentAnnotationLoader` (https://github.com/xwiki/xwiki-commons/pull/265/files#diff-7d816372e1d34e954cccbbf7b7d54f21181f1c92efc8bcd4248ecec56d62fab6)? I basically kept same implementation but I just avoid one map to directly use the priority info from the descriptor. So in theory, there should not be any performance change in there.

Reply:
In theory, it should be possible to store the component descriptors in a `ConcurrentSkipListMap` sorted by priority. Then you would only need to merge the already sorted lists when getting the components from the parent. However, I agree that most likely the sorting here won't have any performance impact.